core/Macros.carp

```diff
@@ -119,6 +119,103 @@
   (defndynamic collect-into [xs f]
     (list 'quote
       (collect-into-internal xs (f) f)))
+
+  (doc empty?
+    "Returns true if the provided data literal is empty, false otherwise.")
+  (defndynamic empty? [xs]
+    (if (= 0 (length xs))
+      true
+      false))
+
+  (defndynamic flip [f]
+    (fn [x y]
+      (f y x)))
+
+  ;; Dynamic.or already exists, but since it's a special form, it can't be passed
+  ;; to higher order functions like reduce. So, we define an alternative here.
+  (defndynamic or-internal [x y]
+    (if x true y))
+
+  (defndynamic curry [f x]
+    (fn [y]
+      (f x y)))
+
+  (doc reduce
+    "Reduces or 'folds' a data literal, such as a list or array, into a single
+    value through successive applications of `f`.")
+  (defndynamic reduce [f x xs]
+    (if (empty? xs)
+      x
+      (reduce f (f x (car xs)) (cdr xs))))
+
+  (defndynamic apply-internal [argument-list acc]
+      (if (empty? argument-list)
+          acc
+          (apply-internal (cdr argument-list) (append acc (list (car
+          argument-list))))))
+
+  (doc apply
+    "Applies the function `f` to the provided argument list, passing each value
+    in the list as an argument to the function.")
+  (defndynamic apply [f argument-list]
+    ;; The let clause here is a tad mysterious at first glance. When passed a
+    ;; standalone function name (i.e. not an application (f x), carp evaluates
+    ;; it into the function's defining form, e.g. foo becomes (defn foo [x] x),
+    ;; commands such as + become (command +) etc.  ;; The binding here accounts
+    ;; for that case, allowing users to pass the function name to apply
+    ;; unquoted.
+    ;;
+    ;; This is necessary for parity across map-internal, zip, and apply.
+    ;; Since map calls its function directly, it takes it as is. Apply, on the
+    ;; other hand, would have to take the quoted form, since it builds a list
+    ;; that serves as the actual application.
+    ;;
+    ;; This is problematic for the user facing map function, since it makes
+    ;; calls to map or zip (which uses apply) as appropriate--unless we support
+    ;; the quoted function name argument in map-internal or the unquoted one in
+    ;; apply, we can't use zip and map-internal in map.
+    (let [function-name (list (cadr f))]
+      (apply-internal argument-list function-name)))
+
+  (defndynamic map-internal [f xs acc]
+    (if (empty? xs)
+        acc
+        (map-internal f (cdr xs) (cons-last (f (car xs)) acc))))
+
+  (defndynamic zip-internal [f forms acc]
+    (if (reduce or-internal false (map-internal empty? forms (list)))
+        acc
+        (zip-internal
+          f
+          (map-internal cdr forms (list))
+          (let [result (list (apply f (map-internal car forms (list))))]
+            (append acc result)))))
+
+  (defndynamic zip [f :rest forms]
```

Comment:
Should we add documentation for this as well?

Reply:
Yep! b2bb8d0